File: Top Interview 150/Binary Tree General/117. Populating Next Right Pointers in Each Node II/117. Populating Next Right Pointers in Each Node II.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root == null)
            return root;
        Queue<Node> kue = new LinkedList<>();
        kue.offer(root);
        root.next = null;
        while(!kue.isEmpty()){
            int n = kue.size();
            for(int i=0; i<n; i++){
                Node poll = kue.poll();
                if(poll != null){
                    if(poll.left != null)
                        kue.offer(poll.left);
                    if(poll.right != null)
                        kue.offer(poll.right);
                    if(n-1 != i)
                        poll.next = kue.peek();
                }
            }
        }
        return root;
    }
}

File: Top Interview 150/Binary Tree General/0117. Populating Next Right Pointers in Each Node II/117. Populating Next Right Pointers in Each Node II.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

# bfs
class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root
        
        lst = deque([root])
        while lst:
            n = len(lst)
            # print(n)
            for i in range(n):
                cur = lst.popleft()
                # get left and right
                if cur:
                    if cur.left:
                        lst.append(cur.left)
                    if cur.right:
                        lst.append(cur.right)
                    # print(cur, lst)
                    if(i != n-1):
                        cur.next = lst[0]
        return root
    
    
    
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        
        cur = root
        prev, head = None, None
        
        while cur:
            while cur:
                if not prev:
                    head = cur.left if cur.left else cur.right
                if cur.left:
                    if prev:
                        prev.next = cur.left
                    prev = cur.left
                if cur.right:
                    if prev:
                        prev.next = cur.right
                    prev = cur.right
                
                cur = cur.next

            prev = None
            cur = head

        return root